### ISSUE TITLE
STRSUB produces invalid UTF-8 character errors

### ISSUE THREAD

**Issue by GreenDog72** [rednex/rgbds#348]
lines such as `db STRSUB("ポ", 1, 1)` produce the error `invalid UTF-8 character` likely due to strsub only getting a single byte from the multibyte character.

**Comment by dbrotz** [comment]
This is similar to #297. The cause in both cases is that the string functions operate in terms of bytes. As a comment in that issue said, there could be compatibility issues if this were changed and there is the question of whether the new unit should be unicode code points or charmap characters.

**Comment by aaaaaa123456789** [comment]
If compatibility is a serious issue, it should be possible to define a new function that works in terms of UTF-8 characters.

**Comment by dbrotz** [comment]
I'd prefer changing the existing functions since using bytes is clearly wrong, but I don't know how much existing code would break. If we add new STRSUB and STRLEN functions, what should they be called? There are no obvious good names. The best I can think of would be USTRSUB/USTRLEN or U8STRSUB/U8STRLEN, which are kind of ugly.

Also, when you say "characters", do you mean code points or grapheme clusters?

**Comment by aaaaaa123456789** [comment]
@dbrotz Grapheme clusters are probably way out of scope — Unicode isn't simple.

**Comment by dbrotz** [comment]
Yeah. Using grapheme clusters would probably introduce too much complexity. Using code points is much simpler, even though it might result in some confusion.

For example, someone might try `STRSUB("g̈", 1, 1)`, not realizing that "g̈" is two code points, and be surprised to find that the resulting substring is "g" rather than "g̈" as they expected.

However, as long as the fact that it works in terms of code points is clearly documented, it shouldn't be a problem.

**Comment by AntonioND** [comment]
This is a bug, ``STRSUB`` should use UTF-8 rather than bytes as it is expected to work with strings, not random array of bytes. However, I'd like to know if someone is already using this in some way that may break before fixing it.

**Comment by dbrotz** [comment]
I searched on GitHub for RGBDS code using STRSUB and every use I found was using on ASCII strings. Since ASCII is a subset of UTF-8, switching to code points instead of bytes won't break any code that was only using ASCII strings.

**Comment by dbrotz** [comment]
I searched on GitHub for RGBDS code using STRSUB and in every case I saw, STRSUB was being used on ASCII strings. ASCII characters are always both 1 byte and 1 unicode code point, so defining STRSUB in terms of either bytes or code points is equivalent when the string is ASCII. We don't know if there is any private code that uses STRSUB on non-ASCII strings, but it seems like a safe change for existing code on GitHub.

**Comment by dbrotz** [comment]
I searched on GitHub for RGBDS code using STRSUB and in every case I saw, STRSUB was being used on ASCII strings. ASCII characters are always both 1 byte and 1 unicode code point, so defining STRSUB in terms of either bytes or code points is equivalent when the string is ASCII.

We don't know if there is any private code that uses STRSUB on non-ASCII strings, but it seems like a safe change for existing public code on GitHub.
